### ISSUE TITLE
Middle API ? How to get this goodness without the strictures of ORM (fixed schema)?

### ISSUE THREAD

**Issue by iraabbott** [pynamodb/PynamoDB#592]
Hi,

I am not sure if this is an issue or that I am just trying to do something a little different.

pynamodb appears to do an excellent job of imitating an ORM table in django. 

However, I am more interested in the schema flexibility of DynamoDB.  I intend to store a bunch of random objects with some order, typing, etc.  i.e. very few common fields and effectively random schema for much of the data.

If I map it to an ORM, I  use a textfield and serialize JSON to it, and then use the fixed fields as searchable to the pk - hey, it maps to the number of local secondary keys with a couple left over for good measure.

The documentation says there is a middle layer of access, between the low level and the models, but while alluded to, this appears to not be documented.  I am hoping that this does what I have started in another form - converting item, query operations, etc to / from python dicts, etc.

If this exists in documentable form, I am game to  help document it.  Otherwise, I may be interested in creating an interface at this level - i.e. pythonification without the strictures of ORM.

Regards,

Ira Abbott

**Comment by ikonst** [comment]
pynamodb can store some untyped data, e.g. using an untyped ListAttribute or MapAttribute, but generally it has to be aware of fields that you wish to use e.g. in conditions. I would extract anything that I can use systematically (e.g. in indexes and conditions) into the schema, and keep "raw data" schema-less in a MapAttribute.

Would such approach work for you? Alternatively, what do you find appealing in pynamodb vs. just using boto3?

**Comment by iraabbott** [comment]
Thanks for taking the time to reply.   I didn't know about the pynamo when I started. boto3 works, for sure, but you almost must have a serialization layer, which was easy enough for the basic operations.   

That's when I ran in tho this, which seems to have both reasonable feature coverage and an active community.  Assuming all of this momentum continues, seems like the place to be.

If I understand ListAttribute and MapAttribute correctly, this is placing data down within a single list or map field.  I may need to think about that.  I intuitively prefer just taking the thing I am using and dumping it in rather than pushing it 'down a layer'.  However, I am starting to get the idea that for the thing I am working on to be portable for backends, I may need to reconsider. I guess it just makes the scan / filter a bit more complicated if the attribute not at the top, but ultimately, all the same stuff has to happen in terms of keeping track of and abstracting the range keys per 'duck' (typed object).
